### DISCUSSION TITLE
Disable AGC in RSP1A?

### DISCUSSION THREAD

**Discussion by jaymot123** [szpajder/dumphfdl#24] [Q&A] [answered]
I'm experimenting with gain elements and trying to figure out what they do. If I specify `--gain-elements IFGR=20,RFGR=2` dumphfdl says `[WARNING] Not updating IFGR gain because AGC is enabled`, so is there a way to disable AGC so my IFGR gain element setting is used?

**Comment by szpajder** [comment]
This message is produced by SDRPlay library, not by dumphfdl.

I'm on RSP1A and libsdrplay API ver 3.07 and it works for me - I don't see
this message and I can set the gain elements by hand.

Please post your full dumphfdl command line, SDRPlay API version and
receiver type.

**Reply by jaymot123** [reply]
From probing the driver with SoapySDRUtil:
 ```
driver=SDRplay
 hardware=RSP1A
 sdrplay_api_api_version=3.070000
 sdrplay_api_hw_version=255
```

`$ dumphfdl --soapysdr driver=sdrplay --gain-elements IFGR=20,RFGR=2 --system-table /home/jay/SDR_Apps/dumphfdl-master/etc/systable.conf --system-table-save /home/jay/SDR_Apps/systable.conf --output decoded:basestation:tcp:address=127.0.0.1,port=20005 --output decoded:text:file:path=/home/jay/hfdl.log --freq-as-squawk --sample-rate 10000000 13264 13270 13276 13303 13312 13315 13321 13324 13342 13351 17916 17919 17928 17922 17934 17958 17901 17967 17985 21934 21937 21931 21949 21928 21997 21990 21982 21955 21997`

**Reply by jaymot123** [reply]
Terminal contents after running that:
```
dumphfdl 1.2.0 (libacars 2.1.3)
driver=sdrplay: computed center frequency: 17630.500 kHz
System table loaded from /home/jay/SDR_Apps/dumphfdl-master/etc/systable.conf
Found device #0:
  driver = sdrplay
  label = SDRplay Dev0 RSP1A 210502FC98
  serial = 210502FC98
[INFO] devIdx: 0
[INFO] SerNo: 210502FC98
[INFO] hwVer: 255
[INFO] rspDuoMode: 0
[INFO] tuner: 1
[INFO] rspDuoSampleFreq: 0.000000
driver=sdrplay: center frequency set to 17630.500 kHz
driver=sdrplay: frequency correction set to 0.00 ppm
[WARNING] Not updating IFGR gain because AGC is enabled
driver=sdrplay: gain element IFGR set to 50.00 dB
driver=sdrplay: gain element RFGR set to 2.00 dB
driver=sdrplay: using antenna RX
driver=sdrplay: using native sample format CS16 (full_scale: 32767.000)
[INFO] Using format CS16.
Connecting to 127.0.0.1:20005...
Connection to 127.0.0.1:20005 established
```

**Comment by szpajder** [comment, marked as answer]
Can you try the latest devel branch? I have pushed https://github.com/szpajder/dumphfdl/commit/2c0f73479d035f471d92697eb74012857fac266b which might fix this (although I'm still unable to reproduce this here on the same receiver).

**Reply by dg9bja** [reply]
I had the same message as jaymot123. Used latest SoapySDR and a RSP1A.
Devel branch works now without this message.

**Reply by dlpwx** [reply]
I no longer have an RSP1A on hand so I can't test this change, but I did experience the same issue as jaymot123 reported here with that device when I was testing it with dumphfdl. 

Additional notes:
- The latest devel branch change works fine with Airspy Mini and dumphfdl. On Airspy Mini `--gain-elements LNA=0,MIX=0,VGA=4` continues modify gain as expected, but with one continuing issue (not specific to devel branch). If I restart host computer with  --gain-elements LNA=0,MIX=0,VGA=4 in the dumphfdl config, it seems the Airspy Mini will start (systemd service starts at boot) with no gain / AGC off. I'll see a really low noise floor, weak signals if any at all. The simple fix (after each reboot) is to remove  --gain-elements LNA=0,MIX=0,VGA=4 from the dumphfdl config, restart dumphfdl with no --gain-elements, then add back  --gain-elements LNA=0,MIX=0,VGA=4 into the config and restart dumphfdl one more time. After that gain operates as expected. I haven't reported this issue because I don't reboot all that frequently and I figured it could be related to local build environment versions of various libraries. Though I have tried different or latest versions of soapysdr soapyairspy gr-osmo etc. So I'm curious if any other Airspy Mini dumphfdl users might find similar behavior.
- For AirspyHF+ Discovery with  dumphfdl, through a bit of searching, it appears soapyairspyhf does not expose any gain controls through soapy. So by design no --gain or --gain-elements usage in dumphfdl config will actually impact gain. That's perfectly fine for these radios based on my tests since they perform very well across a broad range of source signal  strengths. Just thought I should note this to save other dumphfdl users time.

```
$ SoapySDRUtil --info
######################################################
##     Soapy SDR -- the SDR abstraction library     ##
######################################################

Lib Version: v0.8.1-g00e0312c
API Version: v0.8.0
ABI Version: v0.8
Install root: /usr/local
Search path:  /usr/local/lib/SoapySDR/modules0.8
Module found: /usr/local/lib/SoapySDR/modules0.8/libairspySupport.so   (0.2.0-411f73e)
Module found: /usr/local/lib/SoapySDR/modules0.8/libairspyhfSupport.so (0.2.0-d682533)
Module found: /usr/local/lib/SoapySDR/modules0.8/libremoteSupport.so   (0.6.0-c09b2f1)
Available factories... airspy, airspyhf, remote
```

**Reply by jaymot123** [reply]
I downloaded the development brand source code and built and installed it earlier today and now gain elements are working, with no warning message about IFGR not being used because of AGC.

```
`dumphfdl --soapysdr driver=sdrplay --gain-elements IFGR=20,RFGR=2  --system-table /home/jay/SDR_Apps/dumphfdl-master/etc/systable.conf --system-table-save /home/jay/SDR_Apps/systable.conf --output decoded:basestation:tcp:address=127.0.0.1,port=20005 --output decoded:text:file:path=/home/jay/hfdl.log --freq-as-squawk --sample-rate 10000000 5451 5502 5508 5514 5529 5538 5541 5544 5547 5583 5589 5622 5652 5655 5720 6529 6535 6559 6565 6589 6619 6661 8825 8834 8843 8885 8886 8894 8912 8921 8927 8936 8939 8942 8948 8957 8977 10027 10060 10063 10066 10075 10081 10084 10087 10093 11306 11312 11318 11321 11327 11348 11354 11384 11387 11184 11306 13264 13270 13276 13303 13312 13315 13321 13324 13342 13351 13354
dumphfdl 1.2.0 (libacars 2.1.3)
driver=sdrplay: computed center frequency: 9402.500 kHz
System table loaded from /home/jay/SDR_Apps/dumphfdl-master/etc/systable.conf
Found device #0:
  driver = sdrplay
  label = SDRplay Dev0 RSP1A 210502FC98
  serial = 210502FC98
[INFO] devIdx: 0
[INFO] SerNo: 210502FC98
[INFO] hwVer: 255
[INFO] rspDuoMode: 0
[INFO] tuner: 1
[INFO] rspDuoSampleFreq: 0.000000
driver=sdrplay: center frequency set to 9402.500 kHz
driver=sdrplay: frequency correction set to 0.00 ppm
driver=sdrplay: auto gain disabled
driver=sdrplay: gain element IFGR set to 20.00 dB
driver=sdrplay: gain element RFGR set to 2.00 dB
driver=sdrplay: using antenna RX
driver=sdrplay: using native sample format CS16 (full_scale: 32767.000)
[INFO] Using format CS16.
Connecting to 127.0.0.1:20005...
Connection to 127.0.0.1:20005 established
`
```
 Thank you.

I'll wait to see if you have any replies you'd like to add and if not, mark this as answered.

**Reply by jaymot123** [reply]
Just installed the development branch and gain elements are working perfectly.  Thank you. 
I'm using IFGR=20 and RFGR=2 to test with, on the 13MHz through 21MHz bands with 10MHz sample rate. 
This gives me more control over the gain settings vs. IF AGC and a setpoint, so I can see if I can optimize my SDRPlay's HFDL performance. At least it's something else that I can play around with.

If you don't have anything further that you want to add I'll mark this as answered.

**Reply by jaymot123** [reply]
> I'm still unable to reproduce this here on the same receiver

I was just thinking about this and wondering something. Could it be because your receiver's hardware version is different than mine?  Mine is version 255 per  `SoapySDRUtil --probe=driver=sdrplay`.

**Comment by szpajder** [comment]
Mine is the same.
